File: modules/item-classifier/ios/ReferenceStore.swift
import Foundation

/// A single reference sprite entry loaded from the binary fingerprint file.
struct SpriteReference {
  let label: String
  let pHash: UInt64
  let histogram: [Float] // 64 bins
}

/// Loads and stores the precomputed sprite fingerprints from the bundled binary file.
final class ReferenceStore {
  static let shared = ReferenceStore()

  private(set) var references: [SpriteReference] = []
  private(set) var isLoaded = false

  private init() {}

  /// Load fingerprints from the bundled binary file.
  /// Binary format: [4B magic][4B version][4B count][4B reserved]
  /// Per entry: [8B pHash][256B histogram (64×f32)][1B labelLen][NB label]
  func load() {
    guard !isLoaded else { return }

    guard let url = Bundle.main.url(forResource: "sprite-fingerprints", withExtension: "bin") else {
      print("[ReferenceStore] sprite-fingerprints.bin not found in bundle")
      return
    }

    guard let data = try? Data(contentsOf: url) else {
      print("[ReferenceStore] Failed to read sprite-fingerprints.bin")
      return
    }

    guard data.count >= 16 else {
      print("[ReferenceStore] File too small")
      return
    }

    // Validate magic
    let magic = String(data: data[0..<4], encoding: .ascii)
    guard magic == "MWFP" else {
      print("[ReferenceStore] Invalid magic: \(magic ?? "nil")")
      return
    }

    let version = readUInt32(data, offset: 4)
    guard version == 1 else {
      print("[ReferenceStore] Unsupported version: \(version)")
      return
    }

    let count = Int(readUInt32(data, offset: 8))

    var entries: [SpriteReference] = []
    entries.reserveCapacity(count)

    var offset = 16 // skip header

    for _ in 0..<count {
      guard offset + 8 + 256 + 1 <= data.count else { break }

      // pHash (8 bytes, UInt64 LE) — use memcpy for safe unaligned read
      let pHash = readUInt64(data, offset: offset)
      offset += 8

      // Histogram (64 × Float32 = 256 bytes)
      var histogram = [Float](repeating: 0, count: 64)
      for i in 0..<64 {
        histogram[i] = readFloat32(data, offset: offset + i * 4)
      }
      offset += 256

      // Label length (1 byte) + label (UTF-8)
      let labelLen = Int(data[offset])
      offset += 1

      guard offset + labelLen <= data.count else { break }
      let label = String(data: data[offset..<(offset + labelLen)], encoding: .utf8) ?? ""
      offset += labelLen

      entries.append(SpriteReference(label: label, pHash: pHash, histogram: histogram))
    }

    self.references = entries
    self.isLoaded = true
    print("[ReferenceStore] Loaded \(entries.count) sprite fingerprints")
  }

  // MARK: - Safe unaligned reads via memcpy

  private func readUInt32(_ data: Data, offset: Int) -> UInt32 {
    var value: UInt32 = 0
    _ = data.withUnsafeBytes { ptr in
      memcpy(&value, ptr.baseAddress!.advanced(by: offset), 4)
    }
    return value
  }

  private func readUInt64(_ data: Data, offset: Int) -> UInt64 {
    var value: UInt64 = 0
    _ = data.withUnsafeBytes { ptr in
      memcpy(&value, ptr.baseAddress!.advanced(by: offset), 8)
    }
    return value
  }

  private func readFloat32(_ data: Data, offset: Int) -> Float {
    var value: Float = 0
    _ = data.withUnsafeBytes { ptr in
      memcpy(&value, ptr.baseAddress!.advanced(by: offset), 4)
    }
    return value
  }
}

File: modules/item-classifier/src/android/reference-store.ts
/**
 * Loads pre-computed sprite fingerprints and embeddings from bundled binary files.
 *
 * Parses:
 * - sprite-fingerprints.bin (pHash + histogram per sprite)
 * - sprite-embeddings.bin (576-dim embedding per sprite)
 *
 * Binary formats match the Node.js build scripts exactly.
 */
import * as FileSystem from "expo-file-system/legacy";
import { Asset } from "expo-asset";
import { readUint64LE } from "./perceptual-hash";

export interface SpriteReference {
  label: string;
  /** 64-bit pHash as [hi32, lo32] — no BigInt in Hermes */
  pHash: [number, number];
  /** 64-bin HSV histogram */
  histogram: Float32Array;
}

export interface SpriteEmbedding {
  label: string;
  /** L2-normalized 576-dim embedding */
  embedding: Float32Array;
}

let fingerprintRefs: SpriteReference[] = [];
let embeddingRefs: SpriteEmbedding[] = [];
let loaded = false;

export function getReferences(): SpriteReference[] {
  return fingerprintRefs;
}

export function getEmbeddings(): SpriteEmbedding[] {
  return embeddingRefs;
}

export function isLoaded(): boolean {
  return loaded;
}

/**
 * Load both binary files from Android assets.
 * Safe to call multiple times — subsequent calls are no-ops.
 */
export async function loadReferences(): Promise<void> {
  if (loaded) return;

  await Promise.all([loadFingerprints(), loadEmbeddings()]);
  loaded = true;

  console.log(
    `[ReferenceStore] Loaded ${fingerprintRefs.length} fingerprints, ${embeddingRefs.length} embeddings`
  );
}

async function readAssetToArrayBuffer(
  assetModule: number
): Promise<ArrayBuffer> {
  const [asset] = await Asset.loadAsync(assetModule);
  if (!asset.localUri) {
    throw new Error("Asset has no localUri");
  }

  const base64 = await FileSystem.readAsStringAsync(asset.localUri, {
    encoding: "base64",
  });

  // Decode base64 to ArrayBuffer
  const binaryStr = atob(base64);
  const bytes = new Uint8Array(binaryStr.length);
  for (let i = 0; i < binaryStr.length; i++) {
    bytes[i] = binaryStr.charCodeAt(i);
  }
  return bytes.buffer;
}

/**
 * Parse sprite-fingerprints.bin
 * Format: [4B magic "MWFP"][4B version][4B count][4B reserved]
 * Per entry: [8B pHash][256B histogram (64×f32)][1B labelLen][NB label]
 */
async function loadFingerprints(): Promise<void> {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const buffer = await readAssetToArrayBuffer(
      require("../../../../data/sprite-fingerprints.bin")
    );
    const view = new DataView(buffer);

    // Validate magic
    const magic = String.fromCharCode(
      view.getUint8(0),
      view.getUint8(1),
      view.getUint8(2),
      view.getUint8(3)
    );
    if (magic !== "MWFP") {
      console.error(`[ReferenceStore] Invalid fingerprints magic: ${magic}`);
      return;
    }

    const version = view.getUint32(4, true);
    if (version !== 1) {
      console.error(`[ReferenceStore] Unsupported fingerprints version: ${version}`);
      return;
    }

    const count = view.getUint32(8, true);
    const entries: SpriteReference[] = [];
    let offset = 16;

    for (let i = 0; i < count; i++) {
      if (offset + 8 + 256 + 1 > buffer.byteLength) break;

      // pHash as [hi, lo]
      const pHash = readUint64LE(view, offset);
      offset += 8;

      // Histogram: 64 × float32
      const histogram = new Float32Array(64);
      for (let j = 0; j < 64; j++) {
        histogram[j] = view.getFloat32(offset + j * 4, true);
      }
      offset += 256;

      // Label
      const labelLen = view.getUint8(offset);
      offset += 1;

      if (offset + labelLen > buffer.byteLength) break;

      const labelBytes = new Uint8Array(buffer, offset, labelLen);
      const label = new TextDecoder().decode(labelBytes);
      offset += labelLen;

      entries.push({ label, pHash, histogram });
    }

    fingerprintRefs = entries;
  } catch (err) {
    console.error("[ReferenceStore] Failed to load fingerprints:", err);
  }
}

/**
 * Parse sprite-embeddings.bin
 * Format: [4B magic "MWFE"][4B version=1][4B count][4B dim=576]
 * Per entry: [dim×4B embedding (float32 LE)][1B labelLen][NB label]
 */
async function loadEmbeddings(): Promise<void> {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const buffer = await readAssetToArrayBuffer(
      require("../../../../data/sprite-embeddings.bin")
    );
    const view = new DataView(buffer);

    const magic = String.fromCharCode(
      view.getUint8(0),
      view.getUint8(1),
      view.getUint8(2),
      view.getUint8(3)
    );
    if (magic !== "MWFE") {
      console.error(`[ReferenceStore] Invalid embeddings magic: ${magic}`);
      return;
    }

    const version = view.getUint32(4, true);
    if (version !== 1) {
      console.error(`[ReferenceStore] Unsupported embeddings version: ${version}`);
      return;
    }

    const count = view.getUint32(8, true);
    const dim = view.getUint32(12, true);

    const entries: SpriteEmbedding[] = [];
    let offset = 16;

    for (let i = 0; i < count; i++) {
      const embeddingByteLen = dim * 4;
      if (offset + embeddingByteLen + 1 > buffer.byteLength) break;

      // Embedding: dim × float32
      const embedding = new Float32Array(dim);
      for (let j = 0; j < dim; j++) {
        embedding[j] = view.getFloat32(offset + j * 4, true);
      }
      offset += embeddingByteLen;

      // Label
      const labelLen = view.getUint8(offset);
      offset += 1;

      if (offset + labelLen > buffer.byteLength) break;

      const labelBytes = new Uint8Array(buffer, offset, labelLen);
      const label = new TextDecoder().decode(labelBytes);
      offset += labelLen;

      entries.push({ label, embedding });
    }

    embeddingRefs = entries;
  } catch (err) {
    console.error("[ReferenceStore] Failed to load embeddings:", err);
  }
}
